citadel_workspace_service/src/kernel/mod.rs

```diff
@@ -478,6 +560,184 @@ fn handle_connection(
     });
 }
 
+async fn handle_group_broadcast(
+    group_broadcast: GroupBroadcast,
+    implicated_cid: u64,
+    server_connection_map: &mut HashMap<u64, Connection>,
+    tcp_connection_map: Arc<Mutex<HashMap<Uuid, UnboundedSender<InternalServiceResponse>>>>,
+) {
+    if let Some(connection) = server_connection_map.get_mut(&implicated_cid) {
+        let response = match group_broadcast {
+            GroupBroadcast::Invitation {
+                sender: peer_cid,
+                key: group_key,
+            } => Some(InternalServiceResponse::GroupInvitation(GroupInvitation {
+                cid: implicated_cid,
+                peer_cid,
+                group_key,
+                request_id: None,
+            })),
+
+            GroupBroadcast::RequestJoin {
+                sender: peer_cid,
+                key: group_key,
+            } => connection
+                .groups
+                .get_mut(&group_key)
+                .map(|_group_connection| {
+                    InternalServiceResponse::GroupJoinRequestReceived(GroupJoinRequestReceived {
+                        cid: implicated_cid,
+                        peer_cid,
+                        group_key,
+                        request_id: None,
+                    })
+                }),
+
+            GroupBroadcast::AcceptMembership { target: _, key: _ } => None,
+
+            GroupBroadcast::DeclineMembership { target: _, key } => Some(
+                InternalServiceResponse::GroupRequestDeclined(GroupRequestDeclined {
+                    cid: implicated_cid,
+                    group_key: key,
+                    request_id: None,
+                }),
+            ),
+
+            GroupBroadcast::Message {
+                sender: peer_cid,
+                key: group_key,
+                message,
+            } => connection
+                .groups
+                .get_mut(&group_key)
+                .map(|_group_connection| {
+                    InternalServiceResponse::GroupMessageReceived(GroupMessageReceived {
+                        cid: implicated_cid,
+                        peer_cid,
+                        message: message.into_buffer().into(),
```

Comment:
Possibly: is this .into() call cheap? I.e., does it perform no copying? If copying is going on, then, we can just match the type for the field `message` to whatever the type `message.into_buffer()` yields (if possible). If no copying is going on, then, the code is fine.
```suggestion
                        message: message.into_buffer(),
```